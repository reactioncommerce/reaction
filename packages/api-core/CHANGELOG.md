# @reactioncommerce/api-core

## 3.0.0

### Major Changes

- [#6740](https://github.com/reactioncommerce/reaction/pull/6740) [`fee4dbe95`](https://github.com/reactioncommerce/reaction/commit/fee4dbe952e557db8ca658dc08283ba6c7343af9) Thanks [@sujithvn](https://github.com/sujithvn)! - Node18 upgrade includes assert json

### Minor Changes

- [#6684](https://github.com/reactioncommerce/reaction/pull/6684) [`315bb97ab`](https://github.com/reactioncommerce/reaction/commit/315bb97abc3e70dcb1a89da8adca5468302b24be) Thanks [@sujithvn](https://github.com/sujithvn)! - Filter feature. This new feature provides a common function that can be used in a new query endpoint to get filtered results from any collection.

- [#6739](https://github.com/reactioncommerce/reaction/pull/6739) [`0b4b3385f`](https://github.com/reactioncommerce/reaction/commit/0b4b3385f218c4085d8d9f641a28a6e567f4c918) Thanks [@vanpho93](https://github.com/vanpho93)! - feat: upgrade mongodb to verion 5.0

- [#6588](https://github.com/reactioncommerce/reaction/pull/6588) [`9d26f0528`](https://github.com/reactioncommerce/reaction/commit/9d26f05286a43d99f4b8a2924b5a0644a5a03ccc) Thanks [@vanpho93](https://github.com/vanpho93)! - feat: upgrade mongodb to 4.4

### Patch Changes

- Updated dependencies [[`a50a7b359`](https://github.com/reactioncommerce/reaction/commit/a50a7b359bbb546b7abab0e0bfed4c5d8b5ad759), [`315bb97ab`](https://github.com/reactioncommerce/reaction/commit/315bb97abc3e70dcb1a89da8adca5468302b24be), [`fee4dbe95`](https://github.com/reactioncommerce/reaction/commit/fee4dbe952e557db8ca658dc08283ba6c7343af9)]:
  - @reactioncommerce/api-utils@2.0.0
