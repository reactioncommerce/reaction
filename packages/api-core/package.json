{
  "name": "@reactioncommerce/api-core",
  "version": "3.0.0",
  "description": "This NPM package provides the `ReactionAPICore` class. Use this to build a NodeJS microservice that is compatible with the Reaction Commerce platform, or to build your main Reaction Commerce API if you don't want to start by forking the `https://github.com/reactioncommerce/reaction` project.",
  "main": "./src/index.js",
  "type": "module",
  "sideEffects": false,
  "engines": {
    "node": ">=14.18.1"
  },
  "homepage": "https://github.com/reactioncommerce/reaction",
  "url": "https://github.com/reactioncommerce/reaction",
  "email": "engineering@reactioncommerce.com",
  "repository": {
    "type": "git",
    "url": "https://github.com/reactioncommerce/reaction.git",
    "directory": "packages/api-core"
  },
  "author": "Reaction Commerce <hello@reactioncommerce.com>",
  "license": "GPL-3.0",
  "bugs": {
    "url": "https://github.com/reactioncommerce/reaction/issues"
  },
  "scripts": {
    "test": "jest --maxWorkers=4",
    "test:unit:watch": "jest --maxWorkers=4 --watch",
    "test:unit:file": "jest --no-cache --maxWorkers=4",
    "test:unit:file:watch": "jest --no-cache --maxWorkers=4 --watch",
    "lint:gql": "graphql-schema-linter ./src/**/*.graphql"
  },
  "keywords": [],
  "graphql-schema-linter": {
    "rules": [
      "arguments-have-descriptions",
      "deprecations-have-a-reason",
      "descriptions-are-capitalized",
      "enum-values-have-descriptions",
      "enum-values-sorted-alphabetically",
      "fields-have-descriptions",
      "input-object-fields-sorted-alphabetically",
      "input-object-values-have-descriptions",
      "relay-connection-types-spec",
      "relay-page-info-spec",
      "type-fields-sorted-alphabetically",
      "types-are-capitalized",
      "types-have-descriptions"
    ]
  },
  "dependencies": {
    "@apollo/federation": "^0.38.1",
    "@apollo/server": "^4.2.1",
    "@graphql-tools/merge": "^8.3.12",
    "@graphql-tools/schema": "^9.0.10",
    "@reactioncommerce/api-utils": "^2.0.0",
    "@reactioncommerce/logger": "^1.1.3",
    "@reactioncommerce/random": "^1.0.2",
    "@reactioncommerce/reaction-error": "^1.0.1",
    "body-parser": "^1.19.0",
    "callsite": "^1.0.0",
    "cors": "^2.8.5",
    "cuid": "^2.1.8",
    "dataloader": "^2.0.0",
    "diehard": "^1.5.2",
    "dotenv": "^16.0.1",
    "envalid": "^6.0.1",
    "express": "^4.17.1",
    "graphql-iso-date": "^3.6.1",
    "graphql-redis-subscriptions": "^2.5.0",
    "graphql-subscriptions": "^2.0.0",
    "graphql-tag": "^2.10.0",
    "graphql-ws": "^5.11.2",
    "ioredis": "^5.2.4",
    "lodash": "^4.17.15",
    "mongodb": "4.4.1",
    "promise-retry": "^1.1.1",
    "simpl-schema": "^1.7.0",
    "ws": "^8.11.0"
  },
  "devDependencies": {
    "@babel/core": "~7.9.0",
    "@babel/preset-env": "~7.9.0",
    "@reactioncommerce/babel-remove-es-create-require": "~1.0.0",
    "@reactioncommerce/data-factory": "~1.0.1",
    "babel-plugin-rewire-exports": "~2.1.0",
    "babel-plugin-transform-es2015-modules-commonjs": "~6.26.2",
    "babel-plugin-transform-import-meta": "~1.0.1",
    "graphql": "^16.6.0"
  },
  "peerDependencies": {
    "graphql": "^16.6.0"
  },
  "publishConfig": {
    "access": "public"
  }
}
