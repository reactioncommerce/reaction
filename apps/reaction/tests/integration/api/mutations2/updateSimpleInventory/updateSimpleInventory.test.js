import importAsString from "@reactioncommerce/api-utils/importAsString.js";
import insertPrimaryShop from "@reactioncommerce/api-utils/tests/insertPrimaryShop.js";
import Factory from "/tests/util/factory.js";
import { importPluginsJSONFile, ReactionTestAPICore } from "@reactioncommerce/api-core";

const catalogItemQuery = importAsString("./catalogItemQuery.graphql");
const simpleInventoryQuery = importAsString("./simpleInventoryQuery.graphql");
const updateSimpleInventoryMutation = importAsString("./updateSimpleInventoryMutation.graphql");

jest.setTimeout(300000);

const internalShopId = "123";
const opaqueShopId = "cmVhY3Rpb24vc2hvcDoxMjM="; // reaction/shop:123
const internalProductId = "product1";
const opaqueProductId = "cmVhY3Rpb24vcHJvZHVjdDpwcm9kdWN0MQ==";
const internalVariantId = "variant1";
const internalOptionId1 = "option1";
const opaqueOptionId1 = "cmVhY3Rpb24vcHJvZHVjdDpvcHRpb24x";
const internalOptionId2 = "option2";
const opaqueOptionId2 = "cmVhY3Rpb24vcHJvZHVjdDpvcHRpb24y";
const shopName = "Test Shop";

const product = Factory.Product.makeOne({
  _id: internalProductId,
  ancestors: [],
  handle: "product1",
  isDeleted: false,
  isVisible: true,
  shopId: internalShopId,
  type: "simple"
});

const variant = Factory.Product.makeOne({
  _id: internalVariantId,
  ancestors: [internalProductId],
  attributeLabel: "Variant",
  isDeleted: false,
  isVisible: true,
  shopId: internalShopId,
  type: "variant"
});

const option1 = Factory.Product.makeOne({
  _id: internalOptionId1,
  ancestors: [internalProductId, internalVariantId],
  attributeLabel: "Option",
  isDeleted: false,
  isVisible: true,
  shopId: internalShopId,
  type: "variant"
});

const option2 = Factory.Product.makeOne({
  _id: internalOptionId2,
  ancestors: [internalProductId, internalVariantId],
  attributeLabel: "Option",
  isDeleted: false,
  isVisible: true,
  shopId: internalShopId,
  type: "variant"
});

const adminGroup = Factory.Group.makeOne({
  _id: "adminGroup",
  createdBy: null,
  name: "admin",
  permissions: ["reaction:legacy:inventory/read", "reaction:legacy:inventory/update"],
  slug: "admin",
  shopId: internalShopId
});

const customerGroup = Factory.Group.makeOne({
  _id: "customerGroup",
  createdBy: null,
  name: "customer",
  permissions: ["customer"],
  slug: "customer",
  shopId: internalShopId
});

const mockAdminAccount = Factory.Account.makeOne({
  groups: [adminGroup._id],
  shopId: internalShopId
});

const mockCustomerAccount = Factory.Account.makeOne({
  groups: [customerGroup._id],
  shopId: internalShopId
});

let testApp;
let getCatalogItem;
let simpleInventory;
let updateSimpleInventory;
beforeAll(async () => {
  testApp = new ReactionTestAPICore();
  const plugins = await importPluginsJSONFile("../../../../../plugins.json", (pluginList) => {
    // Remove the `files` plugin when testing. Avoids lots of errors.
    delete pluginList.files;

    return pluginList;
  });
  await testApp.reactionNodeApp.registerPlugins(plugins);
  await testApp.start();

  await insertPrimaryShop(testApp.context, { _id: internalShopId, name: shopName });

  await testApp.collections.Products.insertOne(product);
  await testApp.collections.Products.insertOne(variant);
  await testApp.collections.Products.insertOne(option1);
  await testApp.collections.Products.insertOne(option2);

  await testApp.publishProducts([internalProductId]);

  await testApp.collections.Groups.insertOne(adminGroup);
  await testApp.collections.Groups.insertOne(customerGroup);

  await testApp.createUserAndAccount(mockAdminAccount);
  await testApp.createUserAndAccount(mockCustomerAccount);

  getCatalogItem = testApp.query(catalogItemQuery);
  simpleInventory = testApp.query(simpleInventoryQuery);
  updateSimpleInventory = testApp.mutate(updateSimpleInventoryMutation);
});

// There is no need to delete any test data from collections because
// testApp.stop() will drop the entire test database. Each integration
// test file gets its own test database.
afterAll(() => testApp.stop());

test("throws access-denied when updating simpleInventory if not an admin", async () => {
  await testApp.setLoggedInUser(mockCustomerAccount);

  try {
    await updateSimpleInventory({
      input: {
        productConfiguration: {
          productId: opaqueProductId,
          productVariantId: opaqueOptionId1
        },
        shopId: opaqueShopId,
        isEnabled: true
      }
    });
  } catch (errors) {
    expect(errors[0]).toMatchSnapshot();
  }
});

test("returns SimpleInventory record", async () => {
  await testApp.setLoggedInUser(mockAdminAccount);

  const mutationResult = await updateSimpleInventory({
    input: {
      productConfiguration: {
        productId: opaqueProductId,
        productVariantId: opaqueOptionId1
      },
      shopId: opaqueShopId,
      isEnabled: true
    }
  });
  expect(mutationResult).toEqual({
    updateSimpleInventory: {
      inventoryInfo: {
        canBackorder: false,
        inventoryInStock: 0,
        inventoryReserved: 0,
        isEnabled: true,
        lowInventoryWarningThreshold: 0
      }
    }
  });

  const mutationResult2 = await updateSimpleInventory({
    input: {
      productConfiguration: {
        productId: opaqueProductId,
        productVariantId: opaqueOptionId1
      },
      shopId: opaqueShopId,
      canBackorder: true,
      inventoryInStock: 20,
      lowInventoryWarningThreshold: 2
    }
  });
  expect(mutationResult2).toEqual({
    updateSimpleInventory: {
      inventoryInfo: {
        canBackorder: true,
        inventoryInStock: 20,
        inventoryReserved: 0,
        isEnabled: true,
        lowInventoryWarningThreshold: 2
      }
    }
  });

  const queryResult = await simpleInventory({
    productConfiguration: {
      productId: opaqueProductId,
      productVariantId: opaqueOptionId1
    },
    shopId: opaqueShopId
  });
  expect(queryResult).toEqual({
    simpleInventory: {
      canBackorder: true,
      inventoryInStock: 20,
      inventoryReserved: 0,
      isEnabled: true,
      lowInventoryWarningThreshold: 2
    }
  });
});

test("when all options are sold out and canBackorder, isBackorder is true in Catalog", async () => {
  await testApp.setLoggedInUser(mockAdminAccount);

  await updateSimpleInventory({
    input: {
      productConfiguration: {
        productId: opaqueProductId,
        productVariantId: opaqueOptionId1
      },
      shopId: opaqueShopId,
      isEnabled: true,
      canBackorder: true,
      inventoryInStock: 0
    }
  });

  await updateSimpleInventory({
    input: {
      productConfiguration: {
        productId: opaqueProductId,
        productVariantId: opaqueOptionId2
      },
      shopId: opaqueShopId,
      isEnabled: true,
      canBackorder: true,
      inventoryInStock: 0
    }
  });

  const queryResult = await getCatalogItem({
    slugOrId: product.handle
  });
  expect(queryResult).toEqual({
    catalogItemProduct: {
      product: {
        isBackorder: true,
        isLowQuantity: true,
        isSoldOut: true,
        variants: [{
          canBackorder: true,
          inventoryAvailableToSell: 0,
          inventoryInStock: 0,
          isBackorder: true,
          isLowQuantity: true,
          isSoldOut: true,
          options: [
            {
              canBackorder: true,
              inventoryAvailableToSell: 0,
              inventoryInStock: 0,
              isBackorder: true,
              isLowQuantity: true,
              isSoldOut: true
            },
            {
              canBackorder: true,
              inventoryAvailableToSell: 0,
              inventoryInStock: 0,
              isBackorder: true,
              isLowQuantity: true,
              isSoldOut: true
            }
          ]
        }]
      }
    }
  });
});

test("when all options are sold out and canBackorder is false, isBackorder is false in Catalog", async () => {
  await testApp.setLoggedInUser(mockAdminAccount);

  await updateSimpleInventory({
    input: {
      productConfiguration: {
        productId: opaqueProductId,
        productVariantId: opaqueOptionId1
      },
      shopId: opaqueShopId,
      isEnabled: true,
      canBackorder: false,
      inventoryInStock: 0
    }
  });

  await updateSimpleInventory({
    input: {
      productConfiguration: {
        productId: opaqueProductId,
        productVariantId: opaqueOptionId2
      },
      shopId: opaqueShopId,
      isEnabled: true,
      canBackorder: false,
      inventoryInStock: 0
    }
  });

  const queryResult = await getCatalogItem({
    slugOrId: product.handle
  });
  expect(queryResult).toEqual({
    catalogItemProduct: {
      product: {
        isBackorder: false,
        isLowQuantity: true,
        isSoldOut: true,
        variants: [{
          canBackorder: false,
          inventoryAvailableToSell: 0,
          inventoryInStock: 0,
          isBackorder: false,
          isLowQuantity: true,
          isSoldOut: true,
          options: [
            {
              canBackorder: false,
              inventoryAvailableToSell: 0,
              inventoryInStock: 0,
              isBackorder: false,
              isLowQuantity: true,
              isSoldOut: true
            },
            {
              canBackorder: false,
              inventoryAvailableToSell: 0,
              inventoryInStock: 0,
              isBackorder: false,
              isLowQuantity: true,
              isSoldOut: true
            }
          ]
        }]
      }
    }
  });
});

test("when one option is backordered, isBackorder is true for product in Catalog", async () => {
  await testApp.setLoggedInUser(mockAdminAccount);

  await updateSimpleInventory({
    input: {
      productConfiguration: {
        productId: opaqueProductId,
        productVariantId: opaqueOptionId1
      },
      shopId: opaqueShopId,
      isEnabled: true,
      canBackorder: false,
      inventoryInStock: 10
    }
  });

  await updateSimpleInventory({
    input: {
      productConfiguration: {
        productId: opaqueProductId,
        productVariantId: opaqueOptionId2
      },
      shopId: opaqueShopId,
      isEnabled: true,
      canBackorder: true,
      inventoryInStock: 0
    }
  });

  const queryResult = await getCatalogItem({
    slugOrId: product.handle
  });
  expect(queryResult).toEqual({
    catalogItemProduct: {
      product: {
        isBackorder: false,
        isLowQuantity: true,
        isSoldOut: false,
        variants: [{
          canBackorder: true,
          inventoryAvailableToSell: 10,
          inventoryInStock: 10,
          isBackorder: false,
          isLowQuantity: true,
          isSoldOut: false,
          options: [
            {
              canBackorder: false,
              inventoryAvailableToSell: 10,
              inventoryInStock: 10,
              isBackorder: false,
              isLowQuantity: false,
              isSoldOut: false
            },
            {
              canBackorder: true,
              inventoryAvailableToSell: 0,
              inventoryInStock: 0,
              isBackorder: true,
              isLowQuantity: true,
              isSoldOut: true
            }
          ]
        }]
      }
    }
  });
});

test("all options available", async () => {
  await testApp.setLoggedInUser(mockAdminAccount);

  await updateSimpleInventory({
    input: {
      productConfiguration: {
        productId: opaqueProductId,
        productVariantId: opaqueOptionId1
      },
      shopId: opaqueShopId,
      isEnabled: true,
      canBackorder: true,
      inventoryInStock: 10
    }
  });

  await updateSimpleInventory({
    input: {
      productConfiguration: {
        productId: opaqueProductId,
        productVariantId: opaqueOptionId2
      },
      shopId: opaqueShopId,
      isEnabled: true,
      canBackorder: true,
      inventoryInStock: 20
    }
  });

  const queryResult = await getCatalogItem({
    slugOrId: product.handle
  });
  expect(queryResult).toEqual({
    catalogItemProduct: {
      product: {
        isBackorder: false,
        isLowQuantity: false,
        isSoldOut: false,
        variants: [{
          canBackorder: true,
          inventoryAvailableToSell: 30,
          inventoryInStock: 30,
          isBackorder: false,
          isLowQuantity: false,
          isSoldOut: false,
          options: [
            {
              canBackorder: true,
              inventoryAvailableToSell: 10,
              inventoryInStock: 10,
              isBackorder: false,
              isLowQuantity: false,
              isSoldOut: false
            },
            {
              canBackorder: true,
              inventoryAvailableToSell: 20,
              inventoryInStock: 20,
              isBackorder: false,
              isLowQuantity: false,
              isSoldOut: false
            }
          ]
        }]
      }
    }
  });
});

test("simple-inventory updates during standard order flow", async () => {
  await testApp.setLoggedInUser(mockAdminAccount);

  await updateSimpleInventory({
    input: {
      productConfiguration: {
        productId: opaqueProductId,
        productVariantId: opaqueOptionId1
      },
      shopId: opaqueShopId,
      isEnabled: true,
      canBackorder: true,
      inventoryInStock: 10
    }
  });

  const order = {
    payments: [],
    shipping: [
      {
        items: [
          {
            productId: internalProductId,
            quantity: 2,
            variantId: internalOptionId1
          }
        ]
      }
    ],
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  };

  await testApp.context.appEvents.emit("afterOrderCreate", { order });

  let result = await simpleInventory({
    productConfiguration: {
      productId: opaqueProductId,
      productVariantId: opaqueOptionId1
    },
    shopId: opaqueShopId
  });
  expect(result).toEqual({
    simpleInventory: {
      canBackorder: true,
      inventoryInStock: 10,
      inventoryReserved: 2,
      isEnabled: true,
      lowInventoryWarningThreshold: 2
    }
  });

  await testApp.context.appEvents.emit("afterOrderApprovePayment", { order });

  result = await simpleInventory({
    productConfiguration: {
      productId: opaqueProductId,
      productVariantId: opaqueOptionId1
    },
    shopId: opaqueShopId
  });
  expect(result).toEqual({
    simpleInventory: {
      canBackorder: true,
      inventoryInStock: 8,
      inventoryReserved: 0,
      isEnabled: true,
      lowInventoryWarningThreshold: 2
    }
  });
});

test("simple-inventory updates when canceling before approve", async () => {
  await testApp.setLoggedInUser(mockAdminAccount);

  await updateSimpleInventory({
    input: {
      productConfiguration: {
        productId: opaqueProductId,
        productVariantId: opaqueOptionId1
      },
      shopId: opaqueShopId,
      isEnabled: true,
      canBackorder: true,
      inventoryInStock: 10
    }
  });

  const order = {
    payments: [
      {
        status: "created"
      }
    ],
    shipping: [
      {
        items: [
          {
            productId: internalProductId,
            quantity: 2,
            variantId: internalOptionId1
          }
        ]
      }
    ],
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  };

  await testApp.context.appEvents.emit("afterOrderCreate", { order });

  let result = await simpleInventory({
    productConfiguration: {
      productId: opaqueProductId,
      productVariantId: opaqueOptionId1
    },
    shopId: opaqueShopId
  });
  expect(result).toEqual({
    simpleInventory: {
      canBackorder: true,
      inventoryInStock: 10,
      inventoryReserved: 2,
      isEnabled: true,
      lowInventoryWarningThreshold: 2
    }
  });

  await testApp.context.appEvents.emit("afterOrderCancel", { order });

  result = await simpleInventory({
    productConfiguration: {
      productId: opaqueProductId,
      productVariantId: opaqueOptionId1
    },
    shopId: opaqueShopId
  });
  expect(result).toEqual({
    simpleInventory: {
      canBackorder: true,
      inventoryInStock: 10,
      inventoryReserved: 0,
      isEnabled: true,
      lowInventoryWarningThreshold: 2
    }
  });
});

test("simple-inventory updates when canceling after approve, do not return to stock", async () => {
  await testApp.setLoggedInUser(mockAdminAccount);

  await updateSimpleInventory({
    input: {
      productConfiguration: {
        productId: opaqueProductId,
        productVariantId: opaqueOptionId1
      },
      shopId: opaqueShopId,
      isEnabled: true,
      canBackorder: true,
      inventoryInStock: 10
    }
  });

  const order = {
    payments: [
      {
        status: "created"
      }
    ],
    shipping: [
      {
        items: [
          {
            productId: internalProductId,
            quantity: 2,
            variantId: internalOptionId1
          }
        ]
      }
    ],
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  };

  await testApp.context.appEvents.emit("afterOrderCreate", { order });

  let result = await simpleInventory({
    productConfiguration: {
      productId: opaqueProductId,
      productVariantId: opaqueOptionId1
    },
    shopId: opaqueShopId
  });
  expect(result).toEqual({
    simpleInventory: {
      canBackorder: true,
      inventoryInStock: 10,
      inventoryReserved: 2,
      isEnabled: true,
      lowInventoryWarningThreshold: 2
    }
  });

  order.payments[0].status = "approved";
  await testApp.context.appEvents.emit("afterOrderApprovePayment", { order });

  result = await simpleInventory({
    productConfiguration: {
      productId: opaqueProductId,
      productVariantId: opaqueOptionId1
    },
    shopId: opaqueShopId
  });
  expect(result).toEqual({
    simpleInventory: {
      canBackorder: true,
      inventoryInStock: 8,
      inventoryReserved: 0,
      isEnabled: true,
      lowInventoryWarningThreshold: 2
    }
  });

  await testApp.context.appEvents.emit("afterOrderCancel", { order, returnToStock: false });

  result = await simpleInventory({
    productConfiguration: {
      productId: opaqueProductId,
      productVariantId: opaqueOptionId1
    },
    shopId: opaqueShopId
  });
  expect(result).toEqual({
    simpleInventory: {
      canBackorder: true,
      inventoryInStock: 8,
      inventoryReserved: 0,
      isEnabled: true,
      lowInventoryWarningThreshold: 2
    }
  });
});

test("simple-inventory updates when canceling after approve, do return to stock", async () => {
  await testApp.setLoggedInUser(mockAdminAccount);

  await updateSimpleInventory({
    input: {
      productConfiguration: {
        productId: opaqueProductId,
        productVariantId: opaqueOptionId1
      },
      shopId: opaqueShopId,
      isEnabled: true,
      canBackorder: true,
      inventoryInStock: 10
    }
  });

  const order = {
    payments: [
      {
        status: "created"
      }
    ],
    shipping: [
      {
        items: [
          {
            productId: internalProductId,
            quantity: 2,
            variantId: internalOptionId1
          }
        ]
      }
    ],
    workflow: {
      status: "new",
      workflow: ["new"]
    }
  };

  await testApp.context.appEvents.emit("afterOrderCreate", { order });

  let result = await simpleInventory({
    productConfiguration: {
      productId: opaqueProductId,
      productVariantId: opaqueOptionId1
    },
    shopId: opaqueShopId
  });
  expect(result).toEqual({
    simpleInventory: {
      canBackorder: true,
      inventoryInStock: 10,
      inventoryReserved: 2,
      isEnabled: true,
      lowInventoryWarningThreshold: 2
    }
  });

  order.payments[0].status = "approved";
  await testApp.context.appEvents.emit("afterOrderApprovePayment", { order });

  result = await simpleInventory({
    productConfiguration: {
      productId: opaqueProductId,
      productVariantId: opaqueOptionId1
    },
    shopId: opaqueShopId
  });
  expect(result).toEqual({
    simpleInventory: {
      canBackorder: true,
      inventoryInStock: 8,
      inventoryReserved: 0,
      isEnabled: true,
      lowInventoryWarningThreshold: 2
    }
  });

  await testApp.context.appEvents.emit("afterOrderCancel", { order, returnToStock: true });

  result = await simpleInventory({
    productConfiguration: {
      productId: opaqueProductId,
      productVariantId: opaqueOptionId1
    },
    shopId: opaqueShopId
  });
  expect(result).toEqual({
    simpleInventory: {
      canBackorder: true,
      inventoryInStock: 10,
      inventoryReserved: 0,
      isEnabled: true,
      lowInventoryWarningThreshold: 2
    }
  });
});
