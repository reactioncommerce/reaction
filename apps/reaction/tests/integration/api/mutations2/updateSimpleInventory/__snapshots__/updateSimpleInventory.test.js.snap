// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws access-denied when updating simpleInventory if not an admin 1`] = `
Object {
  "extensions": Object {
    "code": "FORBIDDEN",
    "exception": Object {
      "details": Object {},
      "error": "access-denied",
      "eventData": Object {},
      "isClientSafe": true,
      "reason": "Access Denied",
    },
  },
  "locations": Array [
    Object {
      "column": 3,
      "line": 2,
    },
  ],
  "message": "Access Denied",
  "path": Array [
    "updateSimpleInventory",
  ],
}
`;
